File: core/internal/bench/bench.go
// Package bench implementa o harness de avaliação do SIFT (ramo Benchmarks do
// mapa). Foca nas FILTER-LEVEL METRICS — determinísticas e sem custo de LLM:
// medem se a camada de descoberta expõe a ferramenta certa antes de o agente
// agir. Também compara o custo em tokens do schema TOON vs JSON.
//
// (As downstream agent metrics — que exigem rodar o LLM — ficam para um modo
// opt-in separado, para não gastar tokens em CI.)
package bench

import (
	"context"
	"encoding/json"
	"fmt"
	"os"

	"sift/internal/gateway"
	"sift/internal/registry"
	"sift/internal/toon"
)

// Task é um caso rotulado: a query do usuário e o path "ouro" esperado.
type Task struct {
	Query      string `json:"query"`
	Gold       string `json:"gold"`
	NeedsRisky bool   `json:"needs_risky"`
}

// LoadTasks lê as tasks de um arquivo JSON.
func LoadTasks(path string) ([]Task, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var tasks []Task
	if err := json.Unmarshal(data, &tasks); err != nil {
		return nil, err
	}
	return tasks, nil
}

// TaskResult é o detalhe por task.
type TaskResult struct {
	Query      string
	Gold       string
	GoldRank   int // posição 1-based no menu; 0 = ausente
	Visible    int
	RiskyShown []string
}

// FilterMetrics agrega as métricas de filtro sobre o conjunto de tasks.
type FilterMetrics struct {
	Tasks                 int
	TopK                  int
	GoldExposure          float64 // % de tasks com o gold visível no top-K
	NoVisibleToolRate     float64 // % de tasks SEM o gold no menu
	AvgVisibleTools       float64 // tamanho médio do menu
	AvgExtraTools         float64 // ferramentas além do gold (ruído)
	MRR                   float64 // mean reciprocal rank do gold
	RiskyExposureRate     float64 // % de tasks com alguma tool de risco no menu
	UnauthorizedRiskyRate float64 // % com tool de risco exposta SEM a task precisar
	PerTask               []TaskResult
}

// RunFilter roda a descoberta para cada task e calcula as métricas de filtro.
func RunFilter(ctx context.Context, gw *gateway.Gateway, tasks []Task, topK int) (FilterMetrics, error) {
	risky := gw.Registry().RiskyPaths()
	m := FilterMetrics{Tasks: len(tasks), TopK: topK}

	var goldHits, sumVisible, sumExtra, riskyTasks, unauthorizedTasks int
	var sumRR float64

	for _, t := range tasks {
		res, err := gw.SearchTools(ctx, t.Query, topK)
		if err != nil {
			return m, err
		}

		tr := TaskResult{Query: t.Query, Gold: t.Gold, Visible: len(res)}
		for i, r := range res {
			if r.Path == t.Gold {
				tr.GoldRank = i + 1
			}
			if risky[r.Path] {
				tr.RiskyShown = append(tr.RiskyShown, r.Path)
			}
		}

		sumVisible += tr.Visible
		if tr.GoldRank > 0 {
			goldHits++
			sumRR += 1.0 / float64(tr.GoldRank)
			sumExtra += tr.Visible - 1
		} else {
			sumExtra += tr.Visible
		}
		if len(tr.RiskyShown) > 0 {
			riskyTasks++
			if !t.NeedsRisky {
				unauthorizedTasks++
			}
		}
		m.PerTask = append(m.PerTask, tr)
	}

	n := float64(len(tasks))
	if n > 0 {
		m.GoldExposure = float64(goldHits) / n
		m.NoVisibleToolRate = float64(len(tasks)-goldHits) / n
		m.AvgVisibleTools = float64(sumVisible) / n
		m.AvgExtraTools = float64(sumExtra) / n
		m.MRR = sumRR / n
		m.RiskyExposureRate = float64(riskyTasks) / n
		m.UnauthorizedRiskyRate = float64(unauthorizedTasks) / n
	}
	return m, nil
}

// TokenReport compara o custo em tokens dos schemas (TOON vs JSON).
type TokenReport struct {
	Functions          int
	ToonTokens         int
	CompactJSONTokens  int // o formato compacto do nosso registry
	VerboseJSONTokens  int // JSON Schema estilo OpenAPI (baseline da indústria)
	ReductionVsCompact float64
	ReductionVsVerbose float64
}

// RunTokens computa a economia de tokens do TOON sobre todas as funções.
func RunTokens(reg registry.Registry) TokenReport {
	var rep TokenReport
	for cn, cat := range reg {
		for sn, svc := range cat.Services {
			for fn, f := range svc.Fns {
				path := cn + "." + sn + "." + fn
				rep.Functions++
				rep.ToonTokens += toon.EstimateTokens(toon.EncodeFunction(path, f))

				compact, _ := json.Marshal(f)
				rep.CompactJSONTokens += toon.EstimateTokens(string(compact))

				rep.VerboseJSONTokens += toon.EstimateTokens(verboseJSONSchema(path, f))
			}
		}
	}
	if rep.CompactJSONTokens > 0 {
		rep.ReductionVsCompact = 1 - float64(rep.ToonTokens)/float64(rep.CompactJSONTokens)
	}
	if rep.VerboseJSONTokens > 0 {
		rep.ReductionVsVerbose = 1 - float64(rep.ToonTokens)/float64(rep.VerboseJSONTokens)
	}
	return rep
}

// verboseJSONSchema reconstrói o JSON Schema estilo OpenAPI/Function-Calling que
// um provedor tradicional enviaria — o baseline contra o qual o TOON economiza.
func verboseJSONSchema(path string, fn registry.Function) string {
	props := map[string]any{}
	var required []string
	for name, compact := range fn.P {
		p := registry.ParseParam(name, compact)
		jt := "string"
		if p.Type == "number" {
			jt = "number"
		}
		prop := map[string]any{"type": jt, "description": p.Desc}
		if p.Default != "" {
			prop["default"] = p.Default
		}
		props[name] = prop
		if p.Required {
			required = append(required, name)
		}
	}
	schema := map[string]any{
		"name":        path,
		"description": fn.D,
		"parameters": map[string]any{
			"type":       "object",
			"properties": props,
			"required":   required,
		},
	}
	b, _ := json.MarshalIndent(schema, "", "  ")
	return string(b)
}

// Format imprime um relatório legível das métricas de filtro.
func (m FilterMetrics) Format() string {
	s := fmt.Sprintf(`FILTER-LEVEL METRICS  (tasks=%d, top_k=%d)
  Gold next-tool exposure : %5.1f%%   (alvo: alto)
  No-visible-tool rate    : %5.1f%%   (alvo: 0%%)
  Average visible tools   : %5.2f    (alvo: baixo)
  Extra tools exposed     : %5.2f    (ruído por task)
  MRR (rank do gold)      : %5.3f    (alvo: ->1.0)
  Risky-tool exposure     : %5.1f%%
  Unauthorized risky      : %5.1f%%   (alvo: 0%%)
`,
		m.Tasks, m.TopK,
		m.GoldExposure*100, m.NoVisibleToolRate*100,
		m.AvgVisibleTools, m.AvgExtraTools, m.MRR,
		m.RiskyExposureRate*100, m.UnauthorizedRiskyRate*100)
	return s
}

// Format imprime o relatório de tokens.
func (r TokenReport) Format() string {
	avgToon := 0
	avgVerbose := 0
	if r.Functions > 0 {
		avgToon = r.ToonTokens / r.Functions
		avgVerbose = r.VerboseJSONTokens / r.Functions
	}
	return fmt.Sprintf(`TOKEN COST  (%d funções, estimativa ~4 chars/token)
  TOON total           : %5d tokens  (~%d/tool)
  JSON compacto total  : %5d tokens
  JSON Schema verboso  : %5d tokens  (~%d/tool, baseline OpenAPI)
  Redução TOON vs compacto : %5.1f%%
  Redução TOON vs verboso  : %5.1f%%
`,
		r.Functions,
		r.ToonTokens, avgToon,
		r.CompactJSONTokens,
		r.VerboseJSONTokens, avgVerbose,
		r.ReductionVsCompact*100, r.ReductionVsVerbose*100)
}

File: src/sift/bench.py
"""Evaluation harness (the Benchmarks branch of the design).

Focuses on FILTER-LEVEL METRICS — deterministic, no LLM cost: do we expose the
right tool before the agent acts? Plus a TOON-vs-JSON token comparison.

Downstream agent metrics (which require running an LLM) are intentionally left
out of the default harness so this stays cheap and CI-friendly.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path

from . import toon
from .registry import Registry, ToolDef, param_dict


@dataclass
class Task:
    query: str
    gold: str
    needs_risky: bool = False


def load_tasks(path: str | Path) -> list[Task]:
    data = json.loads(Path(path).read_text(encoding="utf-8"))
    return [Task(t["query"], t["gold"], bool(t.get("needs_risky", False))) for t in data]


@dataclass
class FilterMetrics:
    tasks: int = 0
    top_k: int = 0
    gold_exposure: float = 0.0          # fraction with gold visible in top-k
    no_visible_tool_rate: float = 0.0   # fraction with gold absent
    avg_visible_tools: float = 0.0
    avg_extra_tools: float = 0.0        # noise per task
    mrr: float = 0.0                    # mean reciprocal rank of gold
    risky_exposure_rate: float = 0.0
    unauthorized_risky_rate: float = 0.0
    misses: list[str] = field(default_factory=list)

    def format(self) -> str:
        return (
            f"FILTER-LEVEL METRICS  (tasks={self.tasks}, top_k={self.top_k})\n"
            f"  Gold next-tool exposure : {self.gold_exposure*100:5.1f}%   (target: high)\n"
            f"  No-visible-tool rate    : {self.no_visible_tool_rate*100:5.1f}%   (target: 0%)\n"
            f"  Average visible tools   : {self.avg_visible_tools:5.2f}    (target: low)\n"
            f"  Extra tools exposed     : {self.avg_extra_tools:5.2f}    (noise/task)\n"
            f"  MRR (gold rank)         : {self.mrr:5.3f}    (target: ->1.0)\n"
            f"  Risky-tool exposure     : {self.risky_exposure_rate*100:5.1f}%\n"
            f"  Unauthorized risky      : {self.unauthorized_risky_rate*100:5.1f}%   (target: 0%)\n"
        )


def run_filter(sift, tasks: list[Task], top_k: int = 5) -> FilterMetrics:
    """Run discovery for each task and compute the filter-level metrics."""
    risky = sift.registry.risky_paths()
    m = FilterMetrics(tasks=len(tasks), top_k=top_k)

    gold_hits = sum_visible = sum_extra = risky_tasks = unauth = 0
    sum_rr = 0.0

    for t in tasks:
        results = sift.search_tools(t.query, top_k)
        visible = len(results)
        sum_visible += visible

        rank = 0
        risky_shown = False
        for i, r in enumerate(results):
            if r.path == t.gold:
                rank = i + 1
            if r.path in risky:
                risky_shown = True

        if rank:
            gold_hits += 1
            sum_rr += 1.0 / rank
            sum_extra += visible - 1
        else:
            sum_extra += visible
            m.misses.append(f"{t.query!r} expected {t.gold}")
        if risky_shown:
            risky_tasks += 1
            if not t.needs_risky:
                unauth += 1

    n = max(1, len(tasks))
    m.gold_exposure = gold_hits / n
    m.no_visible_tool_rate = (len(tasks) - gold_hits) / n
    m.avg_visible_tools = sum_visible / n
    m.avg_extra_tools = sum_extra / n
    m.mrr = sum_rr / n
    m.risky_exposure_rate = risky_tasks / n
    m.unauthorized_risky_rate = unauth / n
    return m


@dataclass
class TokenReport:
    functions: int = 0
    toon_tokens: int = 0
    compact_json_tokens: int = 0
    verbose_json_tokens: int = 0
    reduction_vs_compact: float = 0.0
    reduction_vs_verbose: float = 0.0

    def format(self) -> str:
        avg_toon = self.toon_tokens // self.functions if self.functions else 0
        avg_verbose = self.verbose_json_tokens // self.functions if self.functions else 0
        return (
            f"TOKEN COST  ({self.functions} functions, ~4 chars/token estimate)\n"
            f"  TOON total           : {self.toon_tokens:5d} tokens  (~{avg_toon}/tool)\n"
            f"  Compact JSON total   : {self.compact_json_tokens:5d} tokens\n"
            f"  Verbose JSON Schema  : {self.verbose_json_tokens:5d} tokens  (~{avg_verbose}/tool, OpenAPI baseline)\n"
            f"  Reduction TOON vs compact : {self.reduction_vs_compact*100:5.1f}%\n"
            f"  Reduction TOON vs verbose : {self.reduction_vs_verbose*100:5.1f}%\n"
        )


def _verbose_json_schema(tool: ToolDef) -> str:
    props, required = {}, []
    for name, p in tool.params.items():
        jt = "number" if p.type == "number" else "string"
        prop = {"type": jt, "description": p.desc}
        if p.default:
            prop["default"] = p.default
        props[name] = prop
        if p.required:
            required.append(name)
    schema = {
        "name": tool.path,
        "description": tool.description,
        "parameters": {"type": "object", "properties": props, "required": required},
    }
    return json.dumps(schema, indent=2, ensure_ascii=False)


def token_report(registry: Registry) -> TokenReport:
    rep = TokenReport()
    for tool in registry.tools():
        rep.functions += 1
        rep.toon_tokens += toon.estimate_tokens(toon.encode_function(tool))
        p_serialisable = {name: param_dict(p) for name, p in tool.params.items()}
        compact = json.dumps({"d": tool.description, "p": p_serialisable, "r": tool.returns},
                             ensure_ascii=False)
        rep.compact_json_tokens += toon.estimate_tokens(compact)
        rep.verbose_json_tokens += toon.estimate_tokens(_verbose_json_schema(tool))
    if rep.compact_json_tokens:
        rep.reduction_vs_compact = 1 - rep.toon_tokens / rep.compact_json_tokens
    if rep.verbose_json_tokens:
        rep.reduction_vs_verbose = 1 - rep.toon_tokens / rep.verbose_json_tokens
    return rep
